File: mockserver-client-dotnet/test/MockServer.Client.Tests/MatcherValueTests.cs
using System.Text.Json;
using FluentAssertions;
using MockServer.Client.Models;
using Xunit;

namespace MockServer.Client.Tests;

/// <summary>
/// Unit tests for the negation escape hatch: a matcher value whose own first character is '!' or '?'
/// is the negation / optional marker on the wire, so a LITERAL such value must be sent in the object
/// form (<c>{"not":false,"value":"!foo"}</c>) rather than a bare string, and both wire forms must
/// decode back. Mirrors the Go client's matcher_value_test.go.
/// </summary>
public class MatcherValueTests
{
    // Options that register HttpRequestConverter, matching the client. MatcherValue itself uses an
    // attribute converter, so its serialization is options-independent, but the HttpRequest-level tests
    // below need the converter to merge the matcher maps onto the wire.
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        Converters = { new HttpRequestConverter() }
    };

    // A value that round-trips through the plain form must STAY a plain string, so existing
    // expectations are byte-identical on the wire.
    [Theory]
    [InlineData("foo", false, false, "\"foo\"")]                 // Literal("foo")
    [InlineData("foo", true, false, "\"!foo\"")]                 // NotLiteral("foo")
    [InlineData("foo", false, true, "\"?foo\"")]                 // OptionalLiteral("foo")
    [InlineData("", false, false, "\"\"")]                       // Literal("")
    [InlineData("foo!bar", false, false, "\"foo!bar\"")]         // marker not at position 0
    [InlineData("!foo", true, false, "\"!!foo\"")]               // NotLiteral("!foo") re-parses to itself
    public void UnambiguousStaysPlainString(string value, bool not, bool optional, string want)
    {
        var matcher = new MatcherValue(value, not, optional);
        JsonSerializer.Serialize(matcher).Should().Be(want);
    }

    // A value the plain form would misread must use the object form, and must carry "not" explicitly
    // even when false so the intent is unmistakable.
    [Theory]
    [InlineData("!foo", false, false, "{\"not\":false,\"value\":\"!foo\"}")]
    [InlineData("?foo", false, false, "{\"not\":false,\"value\":\"?foo\"}")]
    [InlineData("?!foo", false, false, "{\"not\":false,\"value\":\"?!foo\"}")]
    [InlineData("!?foo", false, false, "{\"not\":false,\"value\":\"!?foo\"}")]
    [InlineData("!foo", false, true, "{\"not\":false,\"optional\":true,\"value\":\"!foo\"}")]
    public void AmbiguousUsesObjectFormWithExplicitNot(string value, bool not, bool optional, string want)
    {
        var matcher = new MatcherValue(value, not, optional);
        JsonSerializer.Serialize(matcher).Should().Be(want);
    }

    // NotLiteral("!foo") serialises to "!!foo", which re-parses to exactly itself, so it is NOT
    // ambiguous and stays the shorter plain string. A naive startsWith('!') implementation would escape
    // this unnecessarily and diverge from the Java/Go clients.
    [Fact]
    public void NotLiteralOfMarkerValueStaysShortPlainString()
    {
        MatcherValue.NotLiteral("!foo").IsAmbiguous().Should().BeFalse();
        JsonSerializer.Serialize(MatcherValue.NotLiteral("!foo")).Should().Be("\"!!foo\"");
    }

    // A blank value must NOT use the object form: the server's object-form reader ignores a blank
    // "value", so escaping one would silently drop the matcher.
    [Theory]
    [InlineData("", false)]
    [InlineData("   ", false)]
    [InlineData("", true)]
    public void BlankNeverUsesObjectForm(string value, bool not)
    {
        var json = JsonSerializer.Serialize(new MatcherValue(value, not));
        json[0].Should().Be('"', "a blank value has no marker to misread and the object form drops it");
    }

    [Theory]
    [InlineData("\"foo\"")]
    [InlineData("\"!foo\"")]
    [InlineData("{\"not\":false,\"value\":\"!foo\"}")]
    [InlineData("{\"not\":true,\"value\":\"!foo\"}")]
    public void RoundTripsBothWireForms(string wire)
    {
        var first = JsonSerializer.Deserialize<MatcherValue>(wire);
        var remarshalled = JsonSerializer.Serialize(first);
        var second = JsonSerializer.Deserialize<MatcherValue>(remarshalled);
        second.Should().Be(first);
    }

    [Fact]
    public void ObjectFormDecodesToLiteralValue()
    {
        var matcher = JsonSerializer.Deserialize<MatcherValue>("{\"not\":false,\"value\":\"!foo\"}");
        matcher.Should().Be(MatcherValue.Literal("!foo"));
        matcher.Not.Should().BeFalse();
        matcher.Value.Should().Be("!foo");
    }

    [Fact]
    public void PlainNegatedStringDecodesToNotLiteral()
    {
        var matcher = JsonSerializer.Deserialize<MatcherValue>("\"!foo\"");
        matcher.Should().Be(MatcherValue.NotLiteral("foo"));
        matcher.Not.Should().BeTrue();
        matcher.Value.Should().Be("foo");
    }

    // The matcher maps must REPLACE their plain counterparts on the wire, using the escaped object form.
    [Fact]
    public void MatcherMapsReplacePlainMapsOnTheWire()
    {
        var request = HttpRequest.Request()
            .WithPath("/x")
            .WithHeaderMatcher("X-Tag", MatcherValue.Literal("!foo"))
            .WithQueryStringParameterMatcher("q", MatcherValue.Literal("!foo"))
            .WithCookieMatcher("ck", MatcherValue.Literal("!foo"))
            .WithPathParameterMatcher("id", MatcherValue.Literal("!foo"))
            .Build();

        var doc = JsonDocument.Parse(JsonSerializer.Serialize(request, JsonOptions)).RootElement;

        // header/query/path parameters carry a list; the escaped value is the object form, not "!foo".
        doc.GetProperty("headers").GetProperty("X-Tag")[0].ValueKind.Should().Be(JsonValueKind.Object);
        doc.GetProperty("headers").GetProperty("X-Tag")[0].GetProperty("value").GetString().Should().Be("!foo");
        doc.GetProperty("queryStringParameters").GetProperty("q")[0].GetProperty("value").GetString().Should().Be("!foo");
        doc.GetProperty("pathParameters").GetProperty("id")[0].GetProperty("value").GetString().Should().Be("!foo");
        // cookies carry a single value.
        doc.GetProperty("cookies").GetProperty("ck").GetProperty("value").GetString().Should().Be("!foo");
    }

    // Plain header entries set through WithHeader before the matcher API is used must be carried over,
    // not dropped, when the matcher map replaces the plain field on the wire. This exercises the
    // migration block the four With*Matcher methods share.
    [Fact]
    public void HeaderMatcherCarriesOverExistingPlainEntries()
    {
        var request = HttpRequest.Request()
            .WithPath("/x")
            .WithHeader("X-Existing", "plain")
            .WithHeaderMatcher("X-New", MatcherValue.Literal("!lit"))
            .Build();

        var headers = JsonDocument.Parse(JsonSerializer.Serialize(request, JsonOptions)).RootElement
            .GetProperty("headers");
        headers.GetProperty("X-Existing")[0].GetString().Should().Be("plain", "the pre-existing plain header must survive the migration");
        headers.GetProperty("X-New")[0].GetProperty("value").GetString().Should().Be("!lit");
    }

    // When a matcher map is set directly alongside a plain map (e.g. on an expectation read back from a
    // retrieve), the matcher map REPLACES the plain field on the wire — the documented behaviour of all
    // four matcher maps.
    [Fact]
    public void PathParameterMatchersReplacePlainPathParametersOnTheWire()
    {
        var request = HttpRequest.Request().WithPath("/x/{id}").Build();
        request.PathParameters = new Dictionary<string, List<JsonElement>>
        {
            ["id"] = new List<JsonElement> { JsonDocument.Parse("\"plain\"").RootElement }
        };
        request.PathParameterMatchers = new Dictionary<string, List<MatcherValue>>
        {
            ["id"] = new List<MatcherValue> { MatcherValue.Literal("!lit") }
        };

        var pp = JsonDocument.Parse(JsonSerializer.Serialize(request, JsonOptions)).RootElement
            .GetProperty("pathParameters");
        pp.GetProperty("id")[0].GetProperty("value").GetString().Should().Be("!lit");
    }

    // A genuinely-negated matcher still serialises to the marker string, so it is byte-identical to the
    // plain API and keeps its current meaning.
    [Fact]
    public void NegatedMatcherStillEmitsMarkerString()
    {
        var request = HttpRequest.Request()
            .WithPath("/x")
            .WithHeaderMatcher("X-Tag", MatcherValue.NotLiteral("foo"))
            .Build();

        var doc = JsonDocument.Parse(JsonSerializer.Serialize(request, JsonOptions)).RootElement;
        var value = doc.GetProperty("headers").GetProperty("X-Tag")[0];
        value.ValueKind.Should().Be(JsonValueKind.String);
        value.GetString().Should().Be("!foo");
    }

    // Plain values set via the matcher API serialise exactly as the plain map would.
    [Fact]
    public void PlainMatcherValuesSerialiseAsPlainStrings()
    {
        var request = HttpRequest.Request()
            .WithPath("/x")
            .WithHeaderMatcher("X-Tag", MatcherValue.Literal("foo"))
            .Build();

        var doc = JsonDocument.Parse(JsonSerializer.Serialize(request, JsonOptions)).RootElement;
        doc.GetProperty("headers").GetProperty("X-Tag")[0].GetString().Should().Be("foo");
    }

    // A request using only the plain maps must serialise identically whether or not the converter is
    // registered — the escape is additive and invisible to existing callers.
    [Fact]
    public void PlainOnlyRequestIsUnchangedByTheConverter()
    {
        var request = HttpRequest.Request().WithPath("/x").WithHeader("X-Tag", "foo").Build();

        var withConverter = JsonSerializer.Serialize(request, JsonOptions);
        var withoutConverter = JsonSerializer.Serialize(request, new JsonSerializerOptions
        {
            DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
        });

        withConverter.Should().Be(withoutConverter);
    }

    // The escape must survive a RETRIEVE, not just a create: an expectation carrying the object form
    // must deserialise and re-serialise back to the object form.
    [Fact]
    public void LiteralMatcherSurvivesRoundTripThroughExpectation()
    {
        const string wire =
            "{\"httpRequest\":{\"path\":\"/x\",\"headers\":{\"X-Tag\":[{\"not\":false,\"value\":\"!foo\"}]}}," +
            "\"httpResponse\":{\"statusCode\":200}}";

        var expectation = JsonSerializer.Deserialize<Expectation>(wire, JsonOptions);
        expectation.Should().NotBeNull();

        // The ambiguous value routes to the matcher map, leaving the plain header map null.
        expectation!.HttpRequest!.HeaderMatchers.Should().ContainKey("X-Tag");
        expectation.HttpRequest.HeaderMatchers!["X-Tag"][0].Should().Be(MatcherValue.Literal("!foo"));
        expectation.HttpRequest.Headers.Should().BeNull();

        var reserialised = JsonDocument.Parse(JsonSerializer.Serialize(expectation, JsonOptions)).RootElement;
        reserialised.GetProperty("httpRequest").GetProperty("headers").GetProperty("X-Tag")[0]
            .GetProperty("value").GetString().Should().Be("!foo");
    }

    // A retrieved expectation whose header values all survive the plain form is mirrored back into the
    // plain map, so ordinary traffic decodes exactly as it did before.
    [Fact]
    public void PlainHeaderValuesMirrorBackIntoThePlainMap()
    {
        const string wire =
            "{\"httpRequest\":{\"path\":\"/x\",\"headers\":{\"X-Tag\":[\"foo\",\"!bar\"]}}," +
            "\"httpResponse\":{\"statusCode\":200}}";

        var expectation = JsonSerializer.Deserialize<Expectation>(wire, JsonOptions);

        expectation!.HttpRequest!.HeaderMatchers.Should().BeNull();
        expectation.HttpRequest.Headers.Should().ContainKey("X-Tag");
        expectation.HttpRequest.Headers!["X-Tag"].Should().BeEquivalentTo(new[] { "foo", "!bar" });
    }
}

File: mockserver-client-rust/tests/matcher_value_tests.rs
//! Pure (de)serialization tests for the negation/optional escape hatch.
//!
//! A matcher value that legitimately starts with `!` or `?` collides with
//! MockServer's plain-string marker syntax: the server strips a leading `!`
//! (negation) or `?` (optional) when reading, so a bare `"!foo"` is read as
//! "anything but foo". [`MatcherValue`] escapes such a value to the object form
//! (`{"not":false,"value":"!foo"}`) so the server reads it verbatim, while every
//! value that already round-trips through the plain form stays byte-identical on
//! the wire.

use mockserver_client::{HttpRequest, MatcherValue};
use serde_json::{json, Value};

// A value that round-trips through the plain form must STAY a plain string, so
// existing expectations are byte-identical on the wire.
#[test]
fn unambiguous_stays_plain_string() {
    let cases: &[(MatcherValue, &str)] = &[
        (MatcherValue::literal("foo"), r#""foo""#),
        (MatcherValue::not_literal("foo"), r#""!foo""#),
        (MatcherValue::optional_literal("foo"), r#""?foo""#),
        (MatcherValue::literal(""), r#""""#),
        // marker not at position 0
        (MatcherValue::literal("foo!bar"), r#""foo!bar""#),
        // re-parses to itself, so no escape is needed
        (MatcherValue::not_literal("!foo"), r#""!!foo""#),
    ];
    for (value, want) in cases {
        let got = serde_json::to_string(value).unwrap();
        assert_eq!(&got, want, "serialize {value:?}");
    }
}

// A value the plain form would misread must use the object form, and must carry
// "not" explicitly even when false so the intent is unmistakable.
#[test]
fn ambiguous_uses_object_form_with_explicit_not() {
    let cases: &[(MatcherValue, Value)] = &[
        (
            MatcherValue::literal("!foo"),
            json!({"not": false, "value": "!foo"}),
        ),
        (
            MatcherValue::literal("?foo"),
            json!({"not": false, "value": "?foo"}),
        ),
        (
            MatcherValue::literal("?!foo"),
            json!({"not": false, "value": "?!foo"}),
        ),
        (
            MatcherValue::literal("!?foo"),
            json!({"not": false, "value": "!?foo"}),
        ),
        (
            MatcherValue::optional_literal("!foo"),
            json!({"not": false, "optional": true, "value": "!foo"}),
        ),
    ];
    for (value, want) in cases {
        let got = serde_json::to_value(value).unwrap();
        assert_eq!(&got, want, "serialize {value:?}");
    }
}

// A blank value must NOT use the object form: the server's object-form reader
// ignores a blank "value", so escaping one would silently drop the matcher.
#[test]
fn blank_never_uses_object_form() {
    let cases: &[(MatcherValue, &str)] = &[
        (MatcherValue::literal(""), ""),
        // whitespace-only is blank too, so the value is dropped from the plain form
        (MatcherValue::literal("   "), ""),
        (MatcherValue::not_literal(""), "!"),
    ];
    for (value, want) in cases {
        let got = serde_json::to_value(value).unwrap();
        let got_str = got
            .as_str()
            .unwrap_or_else(|| panic!("serialize {value:?} = {got}, want a plain string"));
        assert_eq!(got_str, *want, "serialize {value:?}");
    }
}

// Both wire forms must decode, and re-encoding must be stable.
#[test]
fn round_trips_both_wire_forms() {
    for wire in [
        json!("foo"),
        json!("!foo"),
        json!({"not": false, "value": "!foo"}),
        json!({"not": true, "value": "!foo"}),
        json!({"not": false, "optional": true, "value": "?foo"}),
    ] {
        let a: MatcherValue = serde_json::from_value(wire.clone()).unwrap();
        let reencoded = serde_json::to_value(&a).unwrap();
        let b: MatcherValue = serde_json::from_value(reencoded).unwrap();
        assert_eq!(a, b, "round-trip {wire}");
    }
}

// `!forbidden` and {"not":true,"value":"forbidden"} are the same matcher; the
// object form normalises to the shorter fixed-point plain string, matching the
// Java NottableString serializer (representation-only, not a mutation).
#[test]
fn negated_object_form_normalises_to_bare_marker() {
    let from_object: MatcherValue =
        serde_json::from_value(json!({"not": true, "value": "forbidden"})).unwrap();
    assert_eq!(from_object, MatcherValue::not_literal("forbidden"));
    assert_eq!(
        serde_json::to_value(&from_object).unwrap(),
        json!("!forbidden")
    );
}

// The plain string builders must keep their current marker-parsing meaning, so
// the new escape is additive rather than a silent behaviour change.
#[test]
fn plain_builders_preserve_marker_meaning() {
    let req = HttpRequest::new()
        .header("X-Neg", "!foo") // still a negation
        .query_param("q", "?maybe") // still optional
        .cookie("ck", "plain");
    let wire = serde_json::to_value(&req).unwrap();
    assert_eq!(wire["headers"]["X-Neg"], json!(["!foo"]));
    assert_eq!(wire["queryStringParameters"]["q"], json!(["?maybe"]));
    assert_eq!(wire["cookies"]["ck"], json!("plain"));
}

// The matcher builders must emit the object form for a literal marker value.
#[test]
fn matcher_builders_emit_object_form_for_literals() {
    let req = HttpRequest::new()
        .path("/x")
        .header_matcher("X-Tag", MatcherValue::literal("!foo"))
        .query_param_matcher("q", MatcherValue::literal("?foo"))
        .cookie_matcher("ck", MatcherValue::literal("!foo"))
        .path_param_matcher("id", MatcherValue::literal("!foo"));
    let wire = serde_json::to_value(&req).unwrap();

    assert_eq!(
        wire["headers"]["X-Tag"][0],
        json!({"not": false, "value": "!foo"})
    );
    assert_eq!(
        wire["queryStringParameters"]["q"][0],
        json!({"not": false, "value": "?foo"})
    );
    assert_eq!(
        wire["cookies"]["ck"],
        json!({"not": false, "value": "!foo"})
    );
    assert_eq!(
        wire["pathParameters"]["id"][0],
        json!({"not": false, "value": "!foo"})
    );
}

// A genuinely negated matcher builder still emits the bare marker string.
#[test]
fn matcher_builder_negation_stays_bare_marker() {
    let req = HttpRequest::new().header_matcher("X-Env", MatcherValue::not_literal("dev"));
    let wire = serde_json::to_value(&req).unwrap();
    assert_eq!(wire["headers"]["X-Env"], json!(["!dev"]));
}

// The escape must survive a decode, not just an encode: the client must decode
// its OWN object-form output back into the literal matcher value.
#[test]
fn object_form_round_trips_through_request_decode() {
    let original = HttpRequest::new()
        .path("/x")
        .header_matcher("X-Tag", MatcherValue::literal("!foo"))
        .cookie_matcher("ck", MatcherValue::literal("!foo"));
    let wire = serde_json::to_value(&original).unwrap();

    let decoded: HttpRequest = serde_json::from_value(wire.clone()).unwrap();
    // An ambiguous value decodes into the additive matcher map, leaving the
    // plain map empty for that field.
    assert_eq!(
        decoded.header_matchers.as_ref().unwrap()["X-Tag"],
        vec![MatcherValue::literal("!foo")]
    );
    assert!(decoded.headers.is_none());
    assert_eq!(
        decoded.cookie_matchers.as_ref().unwrap()["ck"],
        MatcherValue::literal("!foo")
    );
    assert!(decoded.cookies.is_none());

    // and re-encoding reproduces the same wire form
    assert_eq!(serde_json::to_value(&decoded).unwrap(), wire);
}

// Mixing the plain and matcher builders for the same field must not drop the
// plain values: the first matcher call migrates them into the matcher map.
#[test]
fn matcher_builder_migrates_existing_plain_values() {
    let req = HttpRequest::new()
        .path("/x")
        .header("X-Plain", "keep-me")
        .header_matcher("X-Tag", MatcherValue::literal("!foo"));
    let wire = serde_json::to_value(&req).unwrap();
    // both the migrated plain header and the escaped literal are on the wire
    assert_eq!(wire["headers"]["X-Plain"], json!(["keep-me"]));
    assert_eq!(
        wire["headers"]["X-Tag"][0],
        json!({"not": false, "value": "!foo"})
    );
}

// path_param_matcher on a key already carrying plain values replaces them with
// the matcher (object) form; document-and-verify the replacement.
#[test]
fn path_param_matcher_replaces_plain_values_for_same_key() {
    let req = HttpRequest::new()
        .path("/x/{id}")
        .path_param("id", "old")
        .path_param_matcher("id", MatcherValue::literal("!new"));
    let wire = serde_json::to_value(&req).unwrap();
    assert_eq!(
        wire["pathParameters"]["id"],
        json!([{"not": false, "value": "!new"}])
    );
}

// Plain-form traffic must decode into plain-string matcher values exactly as
// before, so nothing about existing behaviour changes on the read path.
#[test]
fn plain_form_decodes_unchanged() {
    let wire = json!({
        "path": "/x",
        "headers": {"X-Tag": ["foo", "!bar"]},
        "cookies": {"ck": "v"}
    });
    let decoded: HttpRequest = serde_json::from_value(wire.clone()).unwrap();
    // Plain-form values decode into the plain map (as strings) and leave the
    // additive matcher map empty, so existing behaviour is unchanged.
    assert_eq!(
        decoded.headers.as_ref().unwrap()["X-Tag"],
        vec!["foo".to_string(), "!bar".to_string()]
    );
    assert!(decoded.header_matchers.is_none());
    assert_eq!(decoded.cookies.as_ref().unwrap()["ck"], "v");
    assert!(decoded.cookie_matchers.is_none());
    // re-encoding is byte-identical to the plain input
    assert_eq!(serde_json::to_value(&decoded).unwrap(), wire);
}
